container/prune_tar.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2015 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This tool build tar files from a container_layer,
+while excluding some files and directories."""
+
+import gflags
+import json
+import sys
+
+from container import archive
+
+gflags.DEFINE_string('output', None, 'The output file, mandatory')
+gflags.MarkFlagAsRequired('output')
+
+gflags.DEFINE_string(
+    'manifest', None, 'JSON manifest of contents to remove from the layer')
+
+gflags.DEFINE_string(
+    'root_directory', '.', 'Default root directory is named "."'
+    'Windows docker images require this be named "Files" instead of "."')
+
+FLAGS = gflags.FLAGS
+
+
+class TarFile(object):
+  """A class to generate a pruned container layer."""
+
+  def __init__(self, output, root_directory, tar, remove_paths):
+    self.output = output
+    self.root_directory = root_directory
+    self.tar = tar
+    self.remove_paths = remove_paths
+
+  def __enter__(self):
+    self.tarfile = archive.TarFileWriter(self.output)
+    return self
+
+  def __exit__(self, t, v, traceback):
+    self.tarfile.close()
+
+  def name_filter(self, path):
+    """Callback for archive.TarFileWriter.add_tar that returns false
+    if the file or directory in remove_paths.
+
+    Args:
+       path: the path to the directory or file in the tar
+    """
+    if path.startswith(self.root_directory + '/'):
+      path = path[len(self.root_directory):]
+
+    for remove_path in self.remove_paths:
+      if remove_path.endswith('/'):
+        if path.startswith(remove_path):
+          return False
+      elif path == remove_path or path.startswith(remove_path + '/'):
+        return False
```

Comment:
Wouldn't just the following cover all cases?

```
if path.startswith(remove_path):
   return False
```

Reply:
Assuming there is the file and folder `/etc/someapp` and `etc/someapp.d` and I want to remove `etc/someapp` only, the simplification would remove too much.

Let me find a way to implement glob so that I don't need this custom logic